internal/pkg/addon/package.go

```diff
@@ -273,8 +281,129 @@ type s3Object struct {
 }
 
 func (a *Addons) uploadAddonAsset(assetPath string, forceZip bool) (s3Object, error) {
+	// make path absolute from wsPath
+	if !path.IsAbs(assetPath) {
+		assetPath = path.Join(a.wsPath, assetPath)
+	}
+
+	info, err := a.fs.Stat(assetPath)
+	if err != nil {
+		return s3Object{}, err
+	}
+
+	var asset asset
+	if forceZip || info.IsDir() {
+		asset, err = a.zipAsset(assetPath)
+	} else {
+		asset, err = a.fileAsset(assetPath)
+	}
+	if err != nil {
+		return s3Object{}, fmt.Errorf("create asset: %w", err)
+	}
+
+	s3Path := artifactpath.AddonAsset(a.wlName, asset.hash)
+
+	url, err := a.uploader.Upload(a.bucket, s3Path, asset.data)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("upload %s to s3 bucket %s: %w", assetPath, a.bucket, err)
+	}
+
+	bucket, key, err := s3.ParseURL(url)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("parse s3 url: %w", err)
+	}
+
+	fmt.Printf("Uploaded %s to s3 at: %s\n", assetPath, s3.Location(bucket, key))
 	return s3Object{
-		Bucket: a.bucket,
-		Key:    "TODO",
+		Bucket: bucket,
+		Key:    key,
+	}, nil
+}
+
+type asset struct {
+	data io.Reader
+	hash string
+}
+
+// zipAsset creates an asset from the directory or file specified by root
+// where the data is the compressed zip archive, and the hash is
+// a hash of each files name, permission, and content.
```

Comment:
Would you like to add an explanation for why it's not the hash of the zip file itself?